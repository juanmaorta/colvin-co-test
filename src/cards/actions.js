export const EDIT_CARD = 'EDIT_CARD'
export const DELETE_CARD = 'DELETE_CARD'
export const FETCH_CARD = 'FETCH_CARD'
export const ADD_CARD = 'ADD_CARD'

export const editCard = (card) => ({
  type: EDIT_CARD,
  card
})

export const deleteCard = (cardId) => ({
  type: DELETE_CARD,
  cardId
})

export const fetchCard = (cardId) => ({
  type: FETCH_CARD,
  cardId
})

export const addCard = (card) => ({
  type: ADD_CARD,
  card
})
