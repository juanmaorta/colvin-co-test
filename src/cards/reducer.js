import * as mocks from '../mocks/cards'
import { ADD_CARD, DELETE_CARD, EDIT_CARD } from './actions'

const defaultState = mocks.cards

export const deleteCard = (cards, cardId) => {
  return cards.filter((card) => {
    return card.id !== cardId && card.parentId !== cardId
  })
}

export const updateCardList = (cards, newCard) => {
  return cards.map((card) => {
    return (card.id !== newCard.id) ? card : newCard
  })
}

export default function (state = defaultState, action) {
  switch (action.type) {
    case ADD_CARD:
      return [
        ...state,
        action.card
      ]
    case DELETE_CARD:
      return deleteCard(state, action.cardId)
    case EDIT_CARD:
      return updateCardList(state, action.card)
    default:
      return state
  }
}
