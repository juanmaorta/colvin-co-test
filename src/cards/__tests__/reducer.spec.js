/* global expect, describe, it, xit */

import reducer, { deleteCard, updateCardList } from '../reducer'
import { ADD_CARD, DELETE_CARD, EDIT_CARD } from '../actions'
import * as stubs from './stubs'

describe('reducer test suite', () => {
  it('should return the default state', () => {
    expect(reducer(undefined, {})).toMatchSnapshot()
  })

  it('should delete a card', () => {
    const action = {
      type: DELETE_CARD,
      cardId: stubs.cardId
    }

    expect(reducer(stubs.cards, action)).toMatchSnapshot()
  })

  it('should add a card', () => {
    const action = {
      type: ADD_CARD,
      card: stubs.newCard
    }

    expect(reducer(stubs.cards, action)).toMatchSnapshot()
  })

  it('should edit a card', () => {
    const editedCard = {
      ...stubs.card,
      text: 'the new text'
    }

    const action = {
      type: EDIT_CARD,
      card: editedCard
    }

    expect(reducer(stubs.cards, action)).toMatchSnapshot()
  })

  describe('reducer functions', () => {
    it('should remove a card from the card list and all its children', () => {
      const expectedCardList = [
        stubs.card1,
        stubs.card2
      ]
      expect(deleteCard(stubs.cards, stubs.card.id)).toEqual(expectedCardList)
    })

    it('should update a card in the card list', () => {
      const updatedCard = {
        ...stubs.card,
        text: 'this is the new text'
      }
      const expectedCardList = [
        updatedCard,
        stubs.card1,
        stubs.card2,
        stubs.childCard
      ]

      expect(updateCardList(stubs.cards, updatedCard)).toEqual(expectedCardList)
    })
  })
})
