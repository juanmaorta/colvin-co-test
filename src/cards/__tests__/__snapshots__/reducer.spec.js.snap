// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer test suite should add a card 1`] = `
Array [
  Object {
    "id": "3c6f4977-c58b-460f-98ad-acd2fae3a2ce",
    "text": "GoldFinger: He's the man, the man with...",
  },
  Object {
    "id": "da1f1278-3f91-4529-8e55-81b4ffd621ac",
    "text": "Instead of coal, Santa now gives the bad children...",
  },
  Object {
    "id": "d50a2288-c93f-448f-8af6-3dc1b7268606",
    "text": "Knock knock. Who's there?",
  },
  Object {
    "id": "2198fcf9-cb27-48e4-befa-aa77da6f1145",
    "parentId": "3c6f4977-c58b-460f-98ad-acd2fae3a2ce",
    "text": "GoldFinger kid",
  },
  Object {
    "id": "b5123b46-1477-4177-be9a-4044cc395a46",
    "text": "to be added",
  },
]
`;

exports[`reducer test suite should delete a card 1`] = `
Array [
  Object {
    "id": "da1f1278-3f91-4529-8e55-81b4ffd621ac",
    "text": "Instead of coal, Santa now gives the bad children...",
  },
  Object {
    "id": "d50a2288-c93f-448f-8af6-3dc1b7268606",
    "text": "Knock knock. Who's there?",
  },
]
`;

exports[`reducer test suite should edit a card 1`] = `
Array [
  Object {
    "id": "3c6f4977-c58b-460f-98ad-acd2fae3a2ce",
    "text": "the new text",
  },
  Object {
    "id": "da1f1278-3f91-4529-8e55-81b4ffd621ac",
    "text": "Instead of coal, Santa now gives the bad children...",
  },
  Object {
    "id": "d50a2288-c93f-448f-8af6-3dc1b7268606",
    "text": "Knock knock. Who's there?",
  },
  Object {
    "id": "2198fcf9-cb27-48e4-befa-aa77da6f1145",
    "parentId": "3c6f4977-c58b-460f-98ad-acd2fae3a2ce",
    "text": "GoldFinger kid",
  },
]
`;

exports[`reducer test suite should return the default state 1`] = `
Array [
  Object {
    "id": "3c6f4977-c58b-460f-98ad-acd2fae3a2ce",
    "text": "GoldFinger: He's the man, the man with...",
  },
  Object {
    "id": "da1f1278-3f91-4529-8e55-81b4ffd621ac",
    "text": "Instead of coal, Santa now gives the bad children...",
  },
  Object {
    "id": "d50a2288-c93f-448f-8af6-3dc1b7268606",
    "text": "Knock knock. Who's there?",
  },
  Object {
    "id": "0f46704f-65c1-47b6-a05b-51eaba442817",
    "parentId": "3c6f4977-c58b-460f-98ad-acd2fae3a2ce",
    "text": "1 ipsum dolor",
  },
  Object {
    "id": "fb681a10-b46d-498b-b35c-2d07001703e8",
    "parentId": "3c6f4977-c58b-460f-98ad-acd2fae3a2ce",
    "text": "2 ipsum dolor",
  },
  Object {
    "id": "b42559cf-efda-42bf-80d9-088976cc697e",
    "parentId": "3c6f4977-c58b-460f-98ad-acd2fae3a2ce",
    "text": "3 ipsum dolor",
  },
]
`;
